pkgs/extension_discovery/README.md

```diff
@@ -0,0 +1,216 @@
+# Package Extension Discovery
+
+A convention to allow other packages to provide extension for your package.
+Including logic for finding extensions that honors this convention.
+
+The convention implemented in this package is that if `foo` provides an
+extension for `<targetPackage>`.
+Then `foo` must contain a config file `extension/<targetPackage>/config.json`.
+This file indicates that `foo` provides and extension for `<targetPackage>`.
+
+If `<targetPackage>` accepts extensions from other packages it must:
+ * Find extensions using `findExtensions('<targetPackage>')` from this package.
+ * Document how extensions are implemented:
+   * What should the contents of the `extension/<targetPackage>/config.json` file be?
+   * Should packages providing extension have a dependency constraint on `<targetPackage>`?
+   * What libraries/assets should packages that provide extension include?
+   * Should packages providing extensions specify a [topic in `pubspec.yaml`][1]
+     for easy discovery on pub.dev.
+
+The `findExtensions(targetPackage, packageConfig: ...)` function will:
+ * Load `.dart_tool/package_config.json` and find all packages that contain a
+   valid JSON file: `extension/<targetPackage>/config.json`.
+ * Provide the package name, location and contents of the `config.json` file,
+   for all detected extensions (aimed at `targetPackage`).
+ * Cache the results for fast loading, comparing modification timestamps to
+   ensure consistent results.
+
+It is the responsibility package that can be extended to validate extensions,
+decide when they should be enabled, and documenting how such extensions are
+created.
+
+
+## Example: Hello World
+
+Imagine that we have a `hello_world` package that defines a single method
+`sayHello(String language)`, along the lines of:
+
+```dart
+void sayHello(String language) {
+  if (language == 'danish') {
+    print('Hej verden');
+  } else {
+    print('Hello world!');
+  }
+}
+```
+
+### Enabling packages to extend `hello_world`
+
+If we wanted to allow other packages to provide additional languages by
+extending the `hello_world` package. Then we could do:
+
+```dart
+import 'package:extension_discovery/extension_discovery.dart';
+
+Future<void> sayHello(String language) async {
+  // Find extensions for the "hello_world" package.
+  // WARNING: This only works when running in JIT-mode, if running in AOT-mode
+  //          you must supply the `packageConfig` argument, and have a local
+  //          `.dart_tool/package_config.json` and `$PUB_CACHE`.
+  final extensions = await findExtensions('hello_world');
```

Comment:
can you provide an example of what this function call would look like in AOT mode when the packageConfig is specified?

Reply:
I'm not sure it easy to make a good example for that.

It's kind of like: `findExtensions('hello_world', packageConfig: Uri.parse('/home/my_user/workspace/myapp/.dart_tool/package_config.json'));`

Where `/home/my_user/workspace/myapp` is the place I did my source checkout and ran `dart pub get`.

---------------------

So in the context of devtools, you'd probably run `findExtensions('hello_world', packageConfig: Uri.parse('path/to/the/users/workspace/.dart_tool/package_config.json'))`.
Where `path/to/the/users/workspace/` is the workspace/app for which the user is running devtools.

Reply:
This was mostly an attempt at pointing to the "Runtime limitations" section.